packages/main/src/Toolbar.ts

```diff
@@ -0,0 +1,497 @@
+			this.distributeItems(contentWidth - containerWidth);
+		} else {
+			this.displayAllItemsIntoBar();
+		}
+
+		this.width = containerWidth;
+
+		if (this.contentWidth !== contentWidth) {
+			this.contentWidth = contentWidth;
+		}
+	}
+
+	storeItemsWidth() {
+		let totalWidth = 0;
+
+		this.movableItems.forEach((item: ToolbarItem) => {
+			const itemWidth = this.getitemWidth(item);
+			const id: string = item._id;
+			totalWidth += itemWidth;
+			this.ITEMS_WIDTH_MAP.set(id, itemWidth);
+		});
+
+		this.ITEMS_WIDTH = totalWidth;
+		this.itemsWidthMeasured = true;
+	}
+
+	distributeItems(overflowSpace = 0) {
+		overflowSpace += this.OVERFLOW_BTN_SIZE;
+
+		this.itemsToBar = [];
+		this.itemsToOverflow = [];
+
+		// distribute items that always overflow
+		this.distributeItemsThatAlwaysOverflow();
+
+		// distribute the rest of the items, based on the available space
+		this.movableItems.reverse().forEach(item => {
+			if (overflowSpace > 0 && item.getAttribute("overflow-priority") !== ToolbarItemOverflowBehavior.NeverOverflow) {
+				this.itemsToOverflow.unshift(item);
+				overflowSpace -= this.getCachedItemWidth(item._id)!;
+			} else {
+				this.itemsToBar.unshift(item);
+			}
+		});
+
+		// If the last bar item is a spacer, force it to the overflow even if there is enough space for it
+		if (this.itemsToBar.length) {
+			const lastItemToBar = this.itemsToBar[this.itemsToBar.length - 1];
+			if (lastItemToBar.ignoreSpace) {
+				const itemBar = this.itemsToBar.pop();
+				if (itemBar) {
+					this.itemsToOverflow.unshift(itemBar);
+				}
+			}
+		}
+	}
+
+	displayAllItemsIntoBar() {
+		this.itemsToOverflow = [];
+
+		// distribute items that always overflow
+		this.distributeItemsThatAlwaysOverflow();
+
+		// distribute items that always overflow
+		this.distributeItemsThatNeverOverflow();
+
+		// distribute the rest of the items into the bar
+		this.itemsToBar = this.movableItems.map((item: ToolbarItem) => item);
+	}
+
+	distributeItemsThatAlwaysOverflow() {
+		this.alwaysOverflowItems.forEach((item: ToolbarItem) => {
+			this.itemsToOverflow.push(item);
+		});
+	}
+
+	distributeItemsThatNeverOverflow() {
+		this.neverOverflowItems.forEach((item: ToolbarItem) => {
+			this.itemsToBar.push(item);
+		});
+	}
+
+	get alwaysOverflowItems() {
+		return this._items.filter((item: ToolbarItem) => item.getAttribute("overflow-priority") === ToolbarItemOverflowBehavior.AlwaysOverflow);
+	}
+
+	get movableItems() {
+		return this._items.filter((item: ToolbarItem) => item.getAttribute("overflow-priority") !== ToolbarItemOverflowBehavior.AlwaysOverflow);
+	}
+
+	get neverOverflowItems() {
+		return this._items.filter((item: ToolbarItem) => item.getAttribute("overflow-priority") === ToolbarItemOverflowBehavior.NeverOverflow);
+	}
+
+	/**
+	 * Event Handlers
+	 */
+	onResize() {
+		if (!this.itemsWidthMeasured) {
+			return;
+		}
+
+		this.resizing = true;
+		this.closeOverflow();
+		this.processOverflowLayout();
+	}
+
+	onCustomItemClick(e: MouseEvent) {
```

Comment:
IMO this name is misleading - aren't all items the same (no standard vs custom AFAIK). Can it be just `onItemClick`? Or the idea was to be opposed to overflow button click?